core/include/gnuradio-4.0/HistoryBuffer.hpp

```diff
@@ -100,23 +113,176 @@ class HistoryBuffer {
+     */
+    template<std::ranges::range Range>
+    constexpr void push_front(const Range& r) noexcept {
+        push_front(std::begin(r), std::end(r));
+    }
+
+    /**
+     * @brief Removes the logical front element (i.e. `operator[](0)`) from the buffer, decreasing the size by one.
+     *
+     * @throws std::out_of_range if the buffer is empty.
+     *
+     * @note
+     * - If you only call `push_back(...)`, `operator[](0)` is the newest element; hence `pop_front()` removes the newest.
+     * - If you only call `push_front(...)`, `operator[](0)` is the oldest element; hence `pop_front()` removes the oldest.
+     * - Mixing both push modes can lead to non-intuitive behavior for front/back usage.
+     */
+    constexpr void pop_front() {
+        if (empty()) {
+            throw std::out_of_range("pop_front() called on empty HistoryBuffer");
+        }
+        if constexpr (N == std::dynamic_extent) {
+            if (++_write_position == _capacity) {
+                _write_position = 0U;
+            }
+        } else {
+            if (_write_position == N - 1U) {
+                _write_position = 0U;
+            } else {
+                ++_write_position;
+            }
+        }
+        --_size;
+    }
+
+    /**
+     * @brief Removes the logical back element (i.e. `operator[](size() - 1)`) from the buffer, decreasing the size by one.
+     *
+     * @throws std::out_of_range if the buffer is empty.
+     *
+     * @note
+     * - If you only call `push_back(...)`, `operator[](size() - 1)` is the oldest element; hence `pop_back()` removes the oldest.
+     * - If you only call `push_front(...)`, `operator[](size() - 1)` is the newest element; hence `pop_back()` removes the newest.
+     * - Mixing both push modes can lead to non-intuitive behavior for front/back usage.
+     */
+    constexpr void pop_back() {
+        if (empty()) {
+            throw std::out_of_range("pop_back() called on empty HistoryBuffer");
+        }
+        // The item at [size()-1] is physically "after" the ring’s start—no need to shift _write_position
+        --_size;
+    }
+
+    constexpr void resize(std::size_t newCapacity)
+    requires(N == std::dynamic_extent)
+    {
+        if (newCapacity == 0) {
+            throw std::out_of_range("new capacity is zero");
+        }
+
+        std::vector<T, Allocator> newBuf(newCapacity * 2);
+
+        const std::size_t copyCount = std::min(_size, newCapacity);
+        const auto        oldFirst  = cbegin();
+        std::copy(oldFirst, oldFirst + static_cast<std::ptrdiff_t>(copyCount), newBuf.begin());                                                        // copy first half
+        std::copy(newBuf.begin(), newBuf.begin() + static_cast<std::ptrdiff_t>(copyCount), newBuf.begin() + static_cast<std::ptrdiff_t>(newCapacity)); // mirror second half
+
+        // update members
+        _buffer         = std::move(newBuf); // destroys old buffer
```

Comment:
`std::swap` might be cheaper here, as newBuf (which would then contain old _buffer data) is anyhow destroyed on end-of-scope.